File: UVA/10696_-_f91/Main.java
import java.util.*;
import java.io.*;

class Main {
  public static int[] table = new int[2 << 19];
  public static int f91(int N) {
    if (table[N] != -1)
      return table[N];
    else {
      if (N <= 100)
        table[N] = f91(f91(N + 11));
      else {
        table[N] = N - 10;
      }
      return table[N];
    }
  }
  public static void main(String[] args) throws Exception {
    Arrays.fill(table, -1);
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    int N;
    while ((N = Integer.parseInt(br.readLine())) != 0) {
      System.out.printf("f91(%d) = %d\n", N, f91(N));
    }
  }
}

File: UVA/10696_-_f91/main.cpp
#include <cstdio>

using namespace std;

static unsigned table[2 << 19];

int f91(int N) {
  if (table[N])
    return table[N];
  else {
    if (N <= 100)
      table[N] = f91(f91(N + 11));
    else
      table[N] = N - 10;
    return table[N];
  }
}

int main() {
  int N;
  while (scanf("%u", &N) && N != 0) {
    printf("f91(%u) = %u\n", N, f91(N));
  }
  return 0;
}
